packages/app/src/cli/services/execute-operation.ts

```diff
@@ -0,0 +1,94 @@
+import {createAdminSessionAsApp, validateSingleOperation} from './graphql/common.js'
+import {OrganizationApp} from '../models/organization.js'
+import {renderSuccess, renderError, renderInfo, renderSingleTask} from '@shopify/cli-kit/node/ui'
+import {outputContent, outputToken, outputResult} from '@shopify/cli-kit/node/output'
+import {AbortError} from '@shopify/cli-kit/node/error'
+import {adminRequestDoc} from '@shopify/cli-kit/node/api/admin'
+import {ClientError} from 'graphql-request'
+import {parse} from 'graphql'
+import {writeFile} from '@shopify/cli-kit/node/fs'
+
+interface ExecuteOperationInput {
+  remoteApp: OrganizationApp
+  storeFqdn: string
+  query: string
+  variables?: string
+  apiVersion?: string
+  outputFile?: string
+}
+
+async function parseVariables(variables?: string): Promise<{[key: string]: unknown} | undefined> {
+  if (!variables) return undefined
+
+  try {
+    return JSON.parse(variables)
+  } catch (error) {
+    const errorMessage = error instanceof Error ? error.message : 'Unknown error'
+    throw new AbortError(
+      outputContent`Invalid JSON in ${outputToken.yellow('--variables')} flag: ${errorMessage}`,
+      'Please provide valid JSON format.',
+    )
+  }
+}
+
+export async function executeOperation(input: ExecuteOperationInput): Promise<void> {
+  const {remoteApp, storeFqdn, query, variables, apiVersion, outputFile} = input
+
+  renderInfo({
```

Comment:
[nit] Now that version is configurable I think it's nice to print it so the user knows what's being executed, can reuse my code from here: https://github.com/Shopify/cli/pull/6684/files#diff-6040bd49d6a47d1cfe82752ae20e0e5502ce7764dcb7ec1b38fa904eea5bbc6fR58-R71

Happy with that being a followup though if makes your life easier :)